### PR TITLE
implement `ignite exec`

### LINKED ISSUE #65: Consider adding `ignite exec [vm] [command]`
It would:
a) use ssh internally
b) require a command, unlike `ssh`

Don't know if we want to do this or not

### PR DESCRIPTION
**Pull request by BenTheElder**
This is a pretty simple / MVP implementation, however it is enough to perform basic exec (over ssh) against a running ignite VM.

TODO:
- inherit out/err/in writers from the cobra command (other commands seem to be ignoring these inputs anyhow...)
- docs (?)
- support for setting environment variables (also we should consider other `docker exec` options? though `-i` already conflicts ...)
- more extensive testing
- consider porting `ignite ssh` to use the same packages

How this works:
- We open an ssh connection to the VM and create a session with golang.org/x/crypto/ssh
- We then request a pseudo tty in the session and wire up stdin / stdout / stderr to the host
- Since ssh requires sending the command as a single string, `args[1:]` are converted to `args[1] + strings.Join(shellEscape(args[2:...], " ")` and run that way (pseudo code)

Additionally, the exit code from the command is preserved if possible.

```console
$ bin/ignite exec centos echo '"hello world"'
"hello world"

$ bin/ignite exec centos -- /bin/bash -c 'echo hello world "$HOSTNAME"'
hello world 50a7f9aa26409ab3
```

Fixes #65

### CODE REVIEW

**Comment by BenTheElder** [comment]
tracking issue: https://github.com/weaveworks/ignite/issues/142

**Comment by alexellis** [comment]
This looks very useful Ben 👍 Out of interest, is it required for KinD to work?

File: cmd/ignite/run/exec.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		return nil, fmt.Errorf("unable to read private key: %v", err)
+	}
+
+	// Create the Signer for this private key.
+	return ssh.ParsePrivateKey(key)
+}
+
+func newSSHConfig(publicKey ssh.Signer, timeout uint32) *ssh.ClientConfig {
+	return &ssh.ClientConfig{
+		User: "root",
```
Comment (alexellis) [inline cmd/ignite/run/exec.go]: Is there a scenario where `root` may not be the user wanted for ssh - VMs with "disable root login"?

File: cmd/ignite/run/exec.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		Do not wrap this error so the caller can check for the exit code
+		If the remote server does not send an exit status, an error of type *ExitMissingError is returned.
+		If the command completes unsuccessfully or is interrupted by a signal, the error is of type *ExitError.
+		Other error types may be returned for I/O problems.
+	*/
+	return session.Run(joinShellCommand(command))
+}
+
+// joinShellCommand joins command parts into a single string safe for passing to sh -c (or SSH)
+func joinShellCommand(command []string) string {
```
Comment (alexellis) [inline cmd/ignite/run/exec.go]: @rgee0 @matipan is this method/package any use to us with faas-cli or ofc-bootstrap?

File: cmd/ignite/run/exec.go (hunk 1 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: I think so, but for the moment this matches `ignite ssh` which also assumes this.

**Comment by BenTheElder** [comment]
> This looks very useful Ben 👍 Out of interest, is it required for KinD to work?

kind requires some way to execute commands, we could implement this in kind some other way (driving SSH? Implementing this logic there?), but I think it's more generally useful to have this in ignite.

---

One other note vs `docker exec`, `docker exec` will treat everything after the container name/ID as part of the command. IE `docker exec centos bash --version` works with `--version` as a argument to bash, not docker vs this PR requiring `ignite exec centos -- bash --version`. 

We could implement this but I think we'd need our own flag processing step, I don't think pflag / cobra alone can support this

File: cmd/ignite/run/exec.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	joined := command[0]
+	if len(command) == 1 {
+		return joined
+	}
+	for _, arg := range command[1:] {
+		// NOTE: we need to escape nested single quotes
```
Comment (BenTheElder) [inline cmd/ignite/run/exec.go]: Slightly stale comment from before using the shellescape package.

File: cmd/ignite/run/exec.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	api "github.com/weaveworks/ignite/pkg/apis/ignite"
+	"github.com/weaveworks/ignite/pkg/constants"
+	"github.com/weaveworks/ignite/pkg/util"
+
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: Remove this blank line and run `make tidy` again :slightly_smiling_face:

File: cmd/ignite/run/exec.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	vm      *api.VM
+	command []string
+}
+
+func (ef *ExecFlags) NewExecOptions(vmMatch string, command ...string) (eo *execOptions, err error) {
+	eo = &execOptions{
+		ExecFlags: ef,
+		command:   command,
+	}
+	eo.vm, err = getVMForMatch(vmMatch)
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: There could be a blank line between the closing brace and `eo.vm` to help readability

File: cmd/ignite/run/exec.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+	// If an external identity file is specified, use it instead of the internal one
+	privKeyFile := eo.IdentityFile
+	if len(privKeyFile) == 0 {
+		privKeyFile = path.Join(eo.vm.ObjectPath(), fmt.Sprintf(constants.VM_SSH_KEY_TEMPLATE, eo.vm.GetUID()))
+		if !util.FileExists(privKeyFile) {
+			return fmt.Errorf("no private key found for VM %q", eo.vm.GetUID())
+		}
+	}
+	signer, err := newSignerForKey(privKeyFile)
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: Here a blank line as well

File: cmd/ignite/run/exec.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	// Create an SSH client, and connect, we will use this to exec
+	config := newSSHConfig(signer, eo.Timeout)
+	client, err := ssh.Dial("tcp", net.JoinHostPort(ipAddrs[0].String(), "22"), config)
+	if err != nil {
+		return fmt.Errorf("failed to dial: %v", err)
+	}
+
+	// run the command, DO NOT wrap this error as the caller can check for the command exit
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: run -> Run

File: cmd/ignite/run/exec.go (hunk 1 above)
Comment (twelho) [reply cmd/ignite/run/exec.go]: The disable root login option default in most distributions prevents password-based authentication, but SSH keys should work just fine. We should also open an issue about allowing user selection in `ssh` and `exec`.

File: cmd/ignite/run/exec.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	defer session.Close()
+
+	// get a pty
+	// TODO: should these be based on the host terminal?
+	// TODO: should we request something other than xterm?
+	// TODO: we should probably configure the terminal modes
+	modes := ssh.TerminalModes{}
+	if err := session.RequestPty("xterm", 80, 40, modes); err != nil {
+		return fmt.Errorf("request for psuedo terminal failed: %v", err)
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: psuedo -> pseudo

File: cmd/ignite/run/exec.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+		return fmt.Errorf("failed to create session: %v", err)
+	}
+	defer session.Close()
+
+	// get a pty
+	// TODO: should these be based on the host terminal?
+	// TODO: should we request something other than xterm?
+	// TODO: we should probably configure the terminal modes
+	modes := ssh.TerminalModes{}
+	if err := session.RequestPty("xterm", 80, 40, modes); err != nil {
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: Defaulting to `xterm` is fine for now, we could later also allow changing this via a flag. Anyways this is non-blocking for this initial implementation.

File: cmd/ignite/run/exec.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	// connect input / output
+	// TODO: these should come from the cobra command instead of hardcoding os.Stderr etc.
+	stderr, err := session.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("failed to connect stderr: %v", err)
+	}
+	go io.Copy(os.Stderr, stderr)
+	stdout, err := session.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("failed to connect stderr: %v", err)
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: stderr -> stdout

File: cmd/ignite/run/exec.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	go io.Copy(os.Stdout, stdout)
+	stdin, err := session.StdinPipe()
+	if err != nil {
+		return fmt.Errorf("failed to connect stderr: %v", err)
```
Comment (twelho) [inline cmd/ignite/run/exec.go]: stderr -> stdin

File: cmd/ignite/run/exec.go (hunk 3 above)
Comment (twelho) [reply cmd/ignite/run/exec.go]: You can go ahead and remove the comment then :+1:

File: cmd/ignite/cmd/vmcmd/exec.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+			}())
+		},
+	}
+
+	addExecFlags(cmd.Flags(), ef)
+	return cmd
+}
+
+func addExecFlags(fs *pflag.FlagSet, ef *run.ExecFlags) {
+	fs.StringVarP(&ef.IdentityFile, "identity", "i", "", "Override the vm's default identity file")
```
[the remark this answers is not in the document]
Comment (BenTheElder) [reply cmd/ignite/cmd/vmcmd/exec.go]: this matches the ssh command's flags exactly at the moment, should we update it there too?

File: cmd/ignite/run/exec.go (hunk 6 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: done

File: cmd/ignite/run/exec.go (hunk 7 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: done

File: cmd/ignite/run/exec.go (hunk 8 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: done

File: cmd/ignite/run/exec.go (hunk 10 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: fixed

File: cmd/ignite/run/exec.go (hunk 11 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: fixed

File: cmd/ignite/run/exec.go (hunk 3 above)
Comment (BenTheElder) [reply cmd/ignite/run/exec.go]: updated to a less stale comment

**Comment by BenTheElder** [comment]
rebased

**Comment by luxas** [comment]
We can follow-up the possible improvements here in an other issue later, but for now, merging this to get us started. Next, integrate with kind :wink:

**Pull request merged**